pkg/tracing/cache.go

```
@@ -0,0 +1,139 @@
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package tracing
+
+import (
+	"errors"
+	"sync"
+
+	"github.com/knative/serving/pkg/tracing/config"
+	zipkin "github.com/openzipkin/zipkin-go"
+	zipkinreporter "github.com/openzipkin/zipkin-go/reporter"
+)
+
+// TracerRef performs reference counting for a tracer's collector. This is needed to
+// allow for closing a collector only after all it's traces are completed.
+type TracerRef struct {
+	Tracer      *zipkin.Tracer
+	reporterRef *refcountedReporter
+}
+
+func (tr *TracerRef) Ref() {
+	if tr.reporterRef != nil {
+		tr.reporterRef.refs.RLock()
+	}
+}
+
+func (tr *TracerRef) Done() {
+	if tr.reporterRef != nil {
+		tr.reporterRef.refs.RUnlock()
+	}
+}
+
+type refcountedReporter struct {
+	refs     sync.RWMutex
+	reporter zipkinreporter.Reporter
+}
+
+// TracerCache manages collector lifecycle and caches the most recently used based
+// on Config. It also allows for the immediate creation of new collectors while
+// outstanding traces exist for the current collector by reference counting.
+//
+// Make sure to call Close() when exiting in order to flush outstanding traces.
+type TracerCache struct {
+	mutex          sync.Mutex
+	reporterRef    *refcountedReporter
+	cfg            *config.Config
+	CreateReporter ReporterFactory
+}
+
+func noopTracer() *TracerRef {
+	tracer, _ := zipkin.NewTracer(nil)
+	return &TracerRef{
+		Tracer:      tracer,
+		reporterRef: nil,
+	}
+}
+
+// NewTracerRef returns a TracerRef for the passed Config. Make sure to call
+// Done on the returned TracerRef when you are done using it. If an invalid
+// config is passed then a previous tracer is returned if it exists along with
+// an error.
+func (tc *TracerCache) NewTracerRef(cfg *config.Config, serviceName, hostPort string) (*TracerRef, error) {
+	tc.mutex.Lock()
+	defer tc.mutex.Unlock()
+
+	if cfg == nil {
+		return noopTracer(), errors.New("Got nil tracer config")
```

This seems weird. As mentioned above, should we surface an error here at all?